File: core/src/main/java/org/treblereel/gwt/three4g/core/Layers.java
package org.treblereel.gwt.three4g.core;

import jsinterop.annotations.JsPackage;
import jsinterop.annotations.JsType;

@JsType(isNative = true, name = "THREE.Layers", namespace = JsPackage.GLOBAL)
public class Layers {
  public double mask;

  public native void disable(double channel);

  public native void disableAll();

  public native void enable(double channel);

  public native void enableAll();

  public native void set(double channel);

  public native boolean test(Layers layers);

  public native void toggle(double channel);
}

File: core/src/main/resources/org/treblereel/gwt/three4g/core/Layers.js
class Layers {

	constructor() {

		/** @type {number} */
		this.mask = 1 | 0;

	}

	/**
	 * 
	 * @param {number} channel 
	 */
	set( channel ) {

		this.mask = 1 << channel | 0;

	}

	/**
	 * @param {number} channel 
	 */
	enable( channel ) {

		this.mask |= 1 << channel | 0;

	}

	enableAll() {

		this.mask = 0xffffffff | 0;

	}

	/**
	 * @param {number} channel 
	 */
	toggle( channel ) {

		this.mask ^= 1 << channel | 0;

	}

	/**
	 * @param {number} channel 
	 */
	disable( channel ) {

		this.mask &= ~ ( 1 << channel | 0 );

	}

	disableAll() {

		this.mask = 0;

	}

	/**
	 * @param {Layers} layers 
	 * @return {boolean}
	 */
	test( layers ) {

		return ( this.mask & layers.mask ) !== 0;

	}

}


export { Layers };
